### ISSUE TITLE
Issue with generic classes

### ISSUE THREAD

**Issue by rubensousa** [evernote/android-state#6]
I started using this library in a new project instead of IcePick and faced this error.

Here's the generic class:

```java
public abstract class LoadingFragment<V extends LoadingContract.View<P>,
        P extends LoadingContract.Presenter<V>> extends BaseFragment<V, P>
        implements LoadingContract.View<P> {

    @State
    boolean loaded = false;

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        StateSaver.saveInstanceState(this, outState);
    }

    @Override
    public void onViewStateRestored(@Nullable Bundle savedInstanceState) {
        super.onViewStateRestored(savedInstanceState);
        StateSaver.restoreInstanceState(this, savedInstanceState);
    }
}
```

After building I get the following error:

```
Error:(12, 68) error: cannot find symbol class V
Error:(12, 71) error: cannot find symbol class P
```

Generated code is:

```java
public class LoadingFragment$$StateSaver<T extends LoadingFragment<V, P>> extends Injector.Object<T> {
  private static final HashMap<String, Bundler<?>> BUNDLERS = new HashMap<String, Bundler<?>>();

  private static final InjectionHelper HELPER = new InjectionHelper("com.github.rubensousa.cepiumapp.ui.base.loading.LoadingFragment$$StateSaver", BUNDLERS);

  static {
  }

  @Override
  @SuppressWarnings("unchecked")
  public void save(T target, Bundle state) {
    HELPER.putBoolean(state, "mLoaded", target.mLoaded);
  }

  @Override
  @SuppressWarnings("unchecked")
  public void restore(T target, Bundle state) {
    target.mLoaded = HELPER.getBoolean(state, "mLoaded");
  }
}
```

IcePick generated this instead:

```java
public class LoadingFragment$$Icepick<T extends LoadingFragment> extends Object<T> {

  private final static Map<String, Bundler<?>> BUNDLERS = new HashMap<String, Bundler<?>>();
  static {
      
  }

  private final static Helper H = new Helper("io.github.rubensousa.euromillions.ui.base.loading.LoadingFragment$$Icepick.", BUNDLERS);

  public void restore(T target, Bundle state) {
    if (state == null) return;
    target.mLoaded = H.getBoolean(state, "mLoaded");
    super.restore(target, state);
  }

  public void save(T target, Bundle state) {
    super.save(target, state);
    H.putBoolean(state, "mLoaded", target.mLoaded);
  }
}
```

This would be solved by removing the generic types in ```T extends LoadingFragment```

**Comment by vRallev** [comment]
Hi, I tried to compile your class, but I'm unable to setup all base classes. Could you provide a sample project? Here's what I have
```java
public interface View <P extends Presenter<View>> {}
```
```java
public class Presenter <V extends View> {}
```
```java
public class BaseFragment<V extends View, P extends Presenter<V>> extends Fragment {}
```
```java
public class LoadingContract {

    public static class Presenter<V extends View<Presenter>> extends com.evernote.android.state.demo.Presenter {

    }

    public interface View <P extends Presenter> extends com.evernote.android.state.demo.View {

    }
}
```
```java
public abstract class LoadingFragment<V extends LoadingContract.View<P>,
        P extends LoadingContract.Presenter<V>> extends BaseFragment<V, P>
        implements LoadingContract.View<P> {

    @State
    boolean loaded = false;

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        StateSaver.saveInstanceState(this, outState);
    }

    @Override
    public void onViewStateRestored(@Nullable Bundle savedInstanceState) {
        super.onViewStateRestored(savedInstanceState);
        StateSaver.restoreInstanceState(this, savedInstanceState);
    }
}
```

But there's an error in the `LoadingFragment`

**Comment by rubensousa** [comment]
Sample project here: https://github.com/rubensousa/StateSaverIssue

**Comment by vRallev** [comment]
Awesome, I was able to reproduce the issue even with a simpler class
```java
public class TestInheritanceGeneric<P> {

    @State
    public int mValue1;

    public P mGenericValue;

    public static class InheritanceLevelGeneric1 <P> extends TestInheritanceGeneric<P> {

    }

    public static class InheritanceLevelGeneric2 extends InheritanceLevelGeneric1<String> {

        @State
        public int mValue2;
    }
}
```

**Comment by rubensousa** [comment]
Awesome! I'll be waiting for a fix

**Comment by vRallev** [comment]
That should be fixed in version `1.0.3`. Please let me know if it works for you!
